File: backend/app/schemas/community.py
import datetime
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel
from app.db.models import (
    BedStatus, CommunityType, MaintenanceStatus, PaymentStatus, RentType,
    RenovationPhase, UnitType,
)


# ---------------------------------------------------------------------------
# Community
# ---------------------------------------------------------------------------

class CommunityCreate(BaseModel):
    community_type: CommunityType
    name: str
    city: str
    province: str = "Alberta"
    operator_id: Optional[int] = None
    has_meal_plan: bool = False
    meal_plan_monthly_cost: Optional[Decimal] = None
    target_occupancy_percent: Optional[Decimal] = None
    description: Optional[str] = None


class CommunityUpdate(BaseModel):
    name: Optional[str] = None
    city: Optional[str] = None
    province: Optional[str] = None
    community_type: Optional[CommunityType] = None
    operator_id: Optional[int] = None
    has_meal_plan: Optional[bool] = None
    meal_plan_monthly_cost: Optional[Decimal] = None
    target_occupancy_percent: Optional[Decimal] = None
    description: Optional[str] = None


class CommunityOut(BaseModel):
    community_id: int
    community_type: CommunityType
    name: str
    city: str
    province: str
    operator_id: Optional[int] = None
    has_meal_plan: bool
    meal_plan_monthly_cost: Optional[Decimal] = None
    target_occupancy_percent: Optional[Decimal] = None
    description: Optional[str] = None

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Unit
# ---------------------------------------------------------------------------

class UnitCreate(BaseModel):
    unit_number: str
    unit_type: UnitType
    bed_count: int
    sqft: Decimal
    floor: Optional[str] = None
    is_legal_suite: bool = False
    notes: Optional[str] = None
    community_id: Optional[int] = None  # optional operational grouping
    monthly_rent: Optional[Decimal] = None  # for by_unit pricing
    bedroom_count: Optional[int] = None  # for by_bedroom pricing
    renovation_phase: RenovationPhase = RenovationPhase.pre_renovation
    development_plan_id: Optional[int] = None  # NULL = baseline, set = plan-specific


class UnitUpdate(BaseModel):
    unit_number: Optional[str] = None
    unit_type: Optional[UnitType] = None
    bed_count: Optional[int] = None
    sqft: Optional[Decimal] = None
    floor: Optional[str] = None
    is_legal_suite: Optional[bool] = None
    is_occupied: Optional[bool] = None
    notes: Optional[str] = None
    community_id: Optional[int] = None
    monthly_rent: Optional[Decimal] = None
    bedroom_count: Optional[int] = None
    renovation_phase: Optional[RenovationPhase] = None
    development_plan_id: Optional[int] = None


class UnitOut(BaseModel):
    unit_id: int
    property_id: int
    community_id: Optional[int] = None
    unit_number: str
    unit_type: UnitType
    bed_count: int
    sqft: Decimal
    floor: Optional[str] = None
    is_legal_suite: bool
    is_occupied: bool
    notes: Optional[str] = None
    monthly_rent: Optional[Decimal] = None
    bedroom_count: Optional[int] = None
    renovation_phase: RenovationPhase = RenovationPhase.pre_renovation
    development_plan_id: Optional[int] = None

    model_config = {"from_attributes": True}


class UnitWithBedsOut(UnitOut):
    """Unit with nested bed details."""
    beds: list["BedOut"] = []

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Bed
# ---------------------------------------------------------------------------

class BedCreate(BaseModel):
    unit_id: Optional[int] = None  # Optional: auto-set from URL path
    bed_label: str
    monthly_rent: Decimal
    rent_type: RentType = RentType.private_pay
    bedroom_number: Optional[int] = None
    is_post_renovation: bool = False
    status: Optional[str] = None  # Optional: defaults to 'available' in model


class BedOut(BaseModel):
    bed_id: int
    unit_id: int
    bed_label: str
    monthly_rent: Decimal
    rent_type: RentType
    status: BedStatus
    bedroom_number: Optional[int] = None
    is_post_renovation: bool = False

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Resident
# ---------------------------------------------------------------------------

class ResidentCreate(BaseModel):
    unit_id: int
    bed_id: Optional[int] = None
    full_name: str
    email: Optional[str] = None
    phone: Optional[str] = None
    bed_number: str
    rent_type: RentType
    move_in_date: datetime.date
    move_out_date: Optional[datetime.date] = None
    enrolled_meal_plan: bool = False


class ResidentOut(BaseModel):
    resident_id: int
    community_id: int
    unit_id: int
    bed_id: Optional[int] = None
    full_name: str
    email: Optional[str] = None
    phone: Optional[str] = None
    bed_number: str
    rent_type: RentType
    move_in_date: datetime.date
    move_out_date: Optional[datetime.date] = None
    enrolled_meal_plan: bool

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Maintenance Request
# ---------------------------------------------------------------------------

class MaintenanceRequestCreate(BaseModel):
    property_id: int
    resident_id: Optional[int] = None
    description: str


class MaintenanceRequestUpdate(BaseModel):
    status: MaintenanceStatus
    resolved_at: Optional[datetime.datetime] = None


class MaintenanceRequestOut(BaseModel):
    request_id: int
    property_id: int
    resident_id: Optional[int] = None
    description: str
    status: MaintenanceStatus
    created_at: datetime.datetime
    resolved_at: Optional[datetime.datetime] = None

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Rent Payment
# ---------------------------------------------------------------------------

class RentPaymentCreate(BaseModel):
    amount: Decimal
    payment_date: datetime.datetime
    period_month: int
    period_year: int
    bed_id: Optional[int] = None
    includes_meal_plan: bool = False


class RentPaymentOut(BaseModel):
    payment_id: int
    resident_id: int
    bed_id: Optional[int] = None
    amount: Decimal
    payment_date: datetime.datetime
    period_month: int
    period_year: int
    status: PaymentStatus
    includes_meal_plan: bool

    model_config = {"from_attributes": True}

File: livingwell-frontend/src/types/community.ts
export type CommunityType = "RecoverWell" | "StudyWell" | "RetireWell";
export type UnitType = "studio" | "1br" | "2br" | "3br" | "suite" | "shared";
export type RentType = "private_pay" | "government_supported" | "shared_room" | "transitional";
export type BedStatus = "available" | "occupied" | "reserved" | "maintenance";
export type MaintenanceStatus = "open" | "in_progress" | "resolved";
export type PaymentStatus = "pending" | "paid" | "overdue";

export interface Community {
  community_id: number;
  community_type: CommunityType;
  name: string;
  city: string;
  province: string;
  operator_id: number | null;
  has_meal_plan: boolean;
  meal_plan_monthly_cost: string | null;
  target_occupancy_percent: string | null;
  description: string | null;
}

export interface Unit {
  unit_id: number;
  community_id: number;
  unit_number: string;
  unit_type: UnitType;
  bed_count: number;
  sqft: string;
  is_occupied: boolean;
}

export interface Bed {
  bed_id: number;
  unit_id: number;
  bed_label: string;
  monthly_rent: string;
  rent_type: RentType;
  status: BedStatus;
}

export interface Resident {
  resident_id: number;
  community_id: number;
  unit_id: number;
  bed_id: number | null;
  full_name: string;
  email: string | null;
  phone: string | null;
  bed_number: string;
  rent_type: RentType;
  move_in_date: string;
  move_out_date: string | null;
  enrolled_meal_plan: boolean;
}

export interface RentPayment {
  payment_id: number;
  resident_id: number;
  bed_id: number | null;
  amount: string;
  payment_date: string;
  period_month: number;
  period_year: number;
  status: PaymentStatus;
  includes_meal_plan: boolean;
}

export interface MaintenanceRequest {
  request_id: number;
  property_id: number;
  resident_id: number | null;
  description: string;
  status: MaintenanceStatus;
  created_at: string;
  resolved_at: string | null;
}
